### DISCUSSION TITLE
socket.lua里的register_protocol 为什么能起效？

### DISCUSSION THREAD

**Discussion by lichangjian** [cloudwu/skynet#1664] [Q&A] [answered]
在socket.lua里有register_protocol代码
```
skynet.register_protocol {
        name = "socket",
	id = skynet.PTYPE_SOCKET,	-- PTYPE_SOCKET = 6
	unpack = driver.unpack,
	dispatch = function (_, _, t, ...)
		socket_message[t](...)
	end
}
```
在模块里引用了socket.lua，当socket_server转发消息给模块的时候，是如何找到socket里注册的protocol的？按我的理解，调用register_protocol，注册的协议是存在 local proto = {} 里的，这是个局部变量。模块与socket.lua 的 proto 不是同一个，在模块里的 proto 应该是找不到socket.lua 里注册的协议才对吧。

**Comment by kevin8990** [comment]
“模块与socket.lua 的 proto 不是同一个”，什么意思，不就一个proto吗

**Reply by lichangjian** [reply]
每次调用 local skynet = require "skynet"，不是都会创建一个 proto 吗？模块 require 了一次，socket.lua 也 require了一次。
请原谅我没写过lua，我理解的require是这样的，是我理解有误？

**Comment by dingdalong** [comment, marked as answer]
你对lua应该不熟悉。
可以去了解一下require相关机制。

**Reply by lichangjian** [reply]
是一个进程只有一个skynet吗？

**Reply by dingdalong** [reply]
你提出的问题，和skynet无关，是lua的特性。
先看看lua的文档，写个最简单的代码，require其他文件，输出一下你想看的数据，就能理解了。

skynet是单进程多线程，其中有多个lua虚拟机，虚拟机之间是独立的，单个虚拟机里面又会包含很多的lua文件。
你现在说的问题在单个虚拟机内部。

**Comment by kevin8990** [comment]
[https://cloudwu.github.io/lua53doc/manual.html#pdf-require](https://cloudwu.github.io/lua53doc/manual.html#pdf-require)，看看require机制
